compiler/rustc_metadata/src/creader.rs

```diff
@@ -184,6 +186,27 @@ impl CStore {
     crate fn has_global_allocator(&self) -> bool {
         self.has_global_allocator
     }
+
+    pub fn report_unused_deps(&self, tcx: TyCtxt<'_>) {
+        // We put the check for the option before the lint_level_at_node call
+        // because the call mutates internal state and introducing it
+        // leads to some ui tests failing.
```

Comment:
Can you say more about these UI test failures? Maybe we should bless them instead?

Reply:
I can't remember exactly which UI test failed but it was unrelated and I didn't want the feature to change stuff unless it's really enabled.

Reply:
I feel uncomfortable landing this without knowing more :)

Reply:
I'll rebase and remove the last commit, then it'll show the failure.

Reply:
@Mark-Simulacrum see the error logs. The entire list is very long. This is an excerpt:

<details>

```

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/reasons-forbidden/reasons-forbidden.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/reasons-forbidden.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/lint/reasons-forbidden.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/reasons-forbidden" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/reasons-forbidden/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error[E0453]: allow(unsafe_code) overruled by outer forbid(unsafe_code)
  --> /checkout/src/test/ui/lint/reasons-forbidden.rs:19:13
   |
LL |     unsafe_code,
   |     ----------- `forbid` level set here
...
LL |     #[allow(unsafe_code)]
   |             ^^^^^^^^^^^ overruled by previous forbid
   |
   = note: our errors & omissions insurance policy doesn't cover unsafe Rust

error: aborting due to 7 previous errors

For more information about this error, try `rustc --explain E0453`.

------------------------------------------


---- [ui] ui/tool_lints.rs stdout ----
diff of stderr:

16	LL | #[warn(foo::bar)]
17	   |        ^^^
18	
-	error: aborting due to 3 previous errors
+	error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
+	  --> $DIR/tool_lints.rs:1:8
+	   |
+	LL | #[warn(foo::bar)]
+	   |        ^^^
+	
+	error: aborting due to 4 previous errors
20	
21	For more information about this error, try `rustc --explain E0710`.
22	


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool_lints/tool_lints.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args tool_lints.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/tool_lints.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool_lints" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/tool_lints/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/tool_lints.rs:1:8
   |
LL | #[warn(foo::bar)]
   |        ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/tool_lints.rs:1:8
   |
LL | #[warn(foo::bar)]
   |        ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/tool_lints.rs:1:8
   |
LL | #[warn(foo::bar)]
   |        ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/tool_lints.rs:1:8
   |
LL | #[warn(foo::bar)]
   |        ^^^

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0710`.

------------------------------------------


---- [ui] ui/unknown-lint-tool-name.rs stdout ----
diff of stderr:

34	LL | #[allow(foo::bar)]
35	   |         ^^^
36	
-	error: aborting due to 6 previous errors
+	error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
+	  --> $DIR/unknown-lint-tool-name.rs:1:9
+	   |
+	LL | #![deny(foo::bar)]
+	   |         ^^^
+	
+	error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
+	  --> $DIR/unknown-lint-tool-name.rs:5:9
+	   |
+	LL | #[allow(foo::bar)]
+	   |         ^^^
+	
+	error: aborting due to 8 previous errors
38	
39	For more information about this error, try `rustc --explain E0710`.
40	


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/unknown-lint-tool-name/unknown-lint-tool-name.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args unknown-lint-tool-name.rs`

error: 1 errors occurred comparing output.
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/unknown-lint-tool-name.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/unknown-lint-tool-name" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/unknown-lint-tool-name/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:1:9
   |
LL | #![deny(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:5:9
   |
LL | #[allow(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:1:9
   |
LL | #![deny(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:5:9
   |
LL | #[allow(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:1:9
   |
LL | #![deny(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:5:9
   |
LL | #[allow(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:1:9
   |
LL | #![deny(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error[E0710]: an unknown tool name found in scoped lint: `foo::bar`
  --> /checkout/src/test/ui/unknown-lint-tool-name.rs:5:9
   |
LL | #[allow(foo::bar)] //~ ERROR an unknown tool name found in scoped lint: `foo::bar`
   |         ^^^

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0710`.

------------------------------------------
```

</details>

Summary of failures:

```
 failures:
    [ui] ui/deduplicate-diagnostics.rs#duplicate
    [ui] ui/error-codes/E0452.rs
    [ui] ui/error-codes/E0453.rs
    [ui] ui/error-codes/E0602.rs
    [ui] ui/feature-gates/feature-gate-lint-reasons.rs
    [ui] ui/feature-gates/feature-gate-unsafe_block_in_unsafe_fn.rs
    [ui] ui/lint/crate_level_only_lint.rs
    [ui] ui/lint/lint-forbid-attr.rs
    [ui] ui/lint/lint-forbid-cmdline.rs
    [ui] ui/lint/lint-malformed.rs
    [ui] ui/lint/lint-unknown-lint-cmdline.rs
    [ui] ui/lint/outer-forbid.rs
    [ui] ui/lint/reasons-erroneous.rs
    [ui] ui/lint/reasons-forbidden.rs
    [ui] ui/tool_lints.rs
    [ui] ui/unknown-lint-tool-name.rs
```

Reply:
Maybe if #78848 is merged it'll be ok.

Reply:
Nope, not fixed by  #78848. Just rebased on top of it and it still fails, just now in the llvm-9 job instead of the llvm-8 one.

Reply:
Ah, ok, I see the cause  -- some things are deduplicated in normal compilation but not here (we force deduplication off for UI testing). Seems fine then.